source/simulator/solver_schemes.cc

```diff
@@ -694,25 +700,30 @@ namespace aspect
                parameters.max_nonlinear_iterations_in_prerefinement)
       :
       parameters.max_nonlinear_iterations;
-    do
+
+    SolverControl nonlinear_solver_control(max_nonlinear_iterations,
+                                           parameters.nonlinear_tolerance,
+                                           true,
+                                           true);
+
+    double relative_residual = std::numeric_limits<double>::max();
+    for (nonlinear_iteration = 0;
+         nonlinear_solver_control.check(nonlinear_iteration, relative_residual) == SolverControl::iterate;
+         ++nonlinear_iteration)
```

Comment:
I think this would be easier to read as a `while` loop:
```
  nonlinear_iteration = 0;
  while (control.check(...) == ...)
  {
    ...
    ++nonlinear_iteration;
  }
```

Reply:
Or maybe even easier, via the `do ... while(...)` construct.